hearpreprocess/util/audio.py

```diff
@@ -3,129 +3,168 @@
 """
 
 import json
-import os
-import subprocess
-from collections import Counter
+from collections import Counter, defaultdict
 from pathlib import Path
-from typing import List, Optional, Union
+from typing import List, Optional, Union, Dict
 
 import numpy as np
-import soundfile as sf
+import ffmpeg
 from tqdm import tqdm
 
 
 def mono_wav_and_fix_duration(in_file: str, out_file: str, duration: float):
     """
     Convert to WAV file and trim to be equal to or less than a specific length
+
+    If the audio is already mono and has the required duration, the step is skipped
+    and a symlink is created to the original file
     """
-    ret = subprocess.call(
-        [
-            "ffmpeg",
-            "-y",
-            "-i",
-            str(in_file),
-            "-filter_complex",
-            f"apad=whole_dur={duration},atrim=end={duration}",
-            "-ac",
-            "1",
-            "-c:a",
-            "pcm_f32le",
-            str(out_file),
-        ],
-        stdout=subprocess.DEVNULL,
-        stderr=subprocess.DEVNULL,
-    )
-    # Make sure the return code is 0 and the command was successful.
-    assert ret == 0, f"ret = {ret}"
-
```

Comment:
I'm not so sure about changing these ffmpeg commnads. They were given to us by Justin Saloman as using the ffmpeg intereface to get speed but the high-quality sox backends. So unless we match these commands exactly (and you show each flag above in the command below), I vote against this change.

Why did you change this?

Reply:
I broke down the command into pieces so that we can select filters to apply depending on the need. If the audio already has the required `duration` or the audio already is `mono` and `.wav` extension, we will create a symlink, otherwise we will add the corresponding things in the chain for ffmpeg to process
```
[
            "ffmpeg",
            "-y",  --> handled by the overwrite_output()
            "-i", --> handled by the input()
            str(in_file),
            "-filter_complex", --> creates a filter chain, this is done by the python command by chaining the steps with the dot
            f"apad=whole_dur={duration},atrim=end={duration}", --> done by the apad and atrim filter
            "-ac", --> this sets the output to mono audio. done in the output
            "1",
            "-c:a", --> This sets the output codec which is done by the acodec filter
            "pcm_f32le",
            str(out_file),
        ],
```
becomes basically this 
```
ffmpeg.input(in_file).audio
.filter("apad", whole_dur=duration)
.filter("atrim", end=duration)
.output(out_file, f="wav", acodec="pcm_f32le", ac=1)
.overwrite_output()
.run(quiet=True)
```

Reply:
The other command which I changed was --

```
[
            "ffmpeg",
            "-y",
            "-i",
            in_file,
            "-af", --> helps to create chained filter which ffmpeg python does
            "aresample=resampler=soxr", --> by the filter aresample
            "-ar", --> in the output
            str(out_sr),
            out_file,
        ],
```
to 
```
  ffmpeg.input(in_file)
  .filter("aresample", resampler="soxr")
  .output(out_file, ar=out_sr)
  .overwrite_output()
  .run(quiet=True)
```